File: loan-calculator/src/main/java/com/jgb/loancalculator/service/PaymentCalculator.java
package com.jgb.loancalculator.service;

import java.math.BigDecimal;
import java.math.RoundingMode;

public class PaymentCalculator {
    public BigDecimal calculate(double amount, double rate, int years) {
        if (rate == 0.0) {
            return calculateWithoutInterest(amount, years);
        } else {
            return calculateWithInterest(amount, rate, years);
        }
    }

    private BigDecimal calculateWithInterest(double amount, double rate, int years) {
        double monthlyRate = rate / 100.0 / 12.0;
        int numberOfPayments = years * 12;
        double payment = (monthlyRate * amount) / (1.0 - Math.pow(1.0 + monthlyRate, -numberOfPayments));
        return toMoney(payment);
    }

    private BigDecimal calculateWithoutInterest(double amount, int years) {
        int numberOfPayments = years * 12;
        return toMoney(amount / numberOfPayments);
    }
    
    private BigDecimal toMoney(double d) {
        BigDecimal bd = new BigDecimal(d);
        return bd.setScale(2, RoundingMode.HALF_UP);
    }
}

File: NetLoanCalculator/Services/PaymentCalculator.cs
using System;

namespace NetLoanCalculator.Services
{
    public class PaymentCalculator
    {
        public Decimal Calculate(double Amount, double Rate, int Years)
        {
            if (Rate == 0.0)
            {
                return CalculateWithoutInterest(Amount, Years);
            }
            else
            {
                return CalculateWithInterest(Amount, Rate, Years);
            }
        }
    
        private Decimal CalculateWithInterest(double Amount, double Rate, int Years)
        {
            double monthlyRate = Rate / 100.0 / 12.0;
            int numberOfPayments = Years * 12;
            double payment = (monthlyRate * Amount) / (1.0 - Math.Pow(1.0 + monthlyRate, -numberOfPayments));
            return ToMoney(payment);
        }

        private Decimal CalculateWithoutInterest(double Amount, int Years)
        {
            int numberOfPayments = Years * 12;
            return ToMoney(Amount / numberOfPayments);
        }
    
        private Decimal ToMoney(double d)
        {
            Decimal bd = new Decimal(d);
            return Decimal.Round(bd, 2, MidpointRounding.AwayFromZero);
        }
    }
}